File: 2023/Day01/day01.jl
A = read("input.txt", String)
lines = split(A, "\n")


digits = [string(x) for x in 1:9]

function get_first_digit(line)
    for l in line
        if l in digits
            return l
        end
    end
end

# Part 1
function solution1()
    result1 = 0
    for line in lines
        letters = split(line, "")
        num = string(get_first_digit(letters), get_first_digit(reverse(letters)))
        result1 = result1 + parse(Int64, num)
    end
    print("Solution 1\n", result1, "\n")
    return result1
end



## Part 2
number_names = ["one","two", "three", "four", "five", "six", "seven", "eight", "nine"]

function replace_front(line)
    indices = [isnothing(findfirst(x, line)) ?  100 : findfirst(x, line)[1] for x in number_names]
    ind_min = minimum(indices)
    newline = deepcopy(line)
    if ind_min < 100
        index = argmin(indices)[1]
        newline = replace(newline, number_names[index] => digits[index], count =1)
    end
    return split(newline, "")
end

function replace_back(line)
    newline = reverse(line)
    indices = [isnothing(findfirst(reverse(x), newline)) ?  100 : findfirst(reverse(x), newline)[1] for x in number_names]
    ind_min = minimum(indices)
    if ind_min < 100
        index = argmin(indices)[1]
        newline = replace(newline, reverse(number_names[index]) => digits[index], count =1)
    end
    return split(newline, "")
end


function solution2()
    result2 = 0
    for line in lines
        num = string(get_first_digit(replace_front(line)), get_first_digit(replace_back(line)))
        result2 = result2 + parse(Int64, num)
    end
    print("Solution 2\n", result2, "\n")
    return result2
end

result1 = solution1()
result2 = solution2()

File: 2023/Day01/day01.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
@author: tobias
"""
import numpy as np


def get_first_digit(line: list[str]) -> str:
    digits = [str(int(x)) for x in np.arange(1,10)]
    for l in line:
        if l in digits:
            return l

switches = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
repls = [str(x) for x in np.arange(1,10,1, dtype=int)]

def replace_front(line: str) -> str:
    indices = [line.find(s) if line.find(s) != -1 else 100 for s in switches]    
    if np.all(np.array(indices)==100):
        1==1
    else:
        change = np.argmin(indices)
        line = line.replace(switches[change], repls[change])
    return line

def replace_back(line: str) -> str:
    newline = line[::-1]
    indices = [newline.find(s[::-1]) if newline.find(s[::-1]) != -1 else 100 for s in switches]
    if np.all(np.array(indices)==100):
        1==1
    else:
        change = np.argmin(indices)
        newline = newline.replace(switches[change][::-1], repls[change])
    return newline

#%% Input reading
with open("input.txt", "r") as f:    
    A = f.read()
lines = A.split()


#%% Part 1
numbers = []
for line in lines:
    letters = list(line)
    num = get_first_digit(letters) +  get_first_digit(letters[::-1])
    numbers.append(int(num))
res1 = sum(numbers)
print(f"Solution 1:\n{res1}")


#%% Part 2
numbers = []
for line in lines:
   number = get_first_digit(replace_front(line)) + get_first_digit(replace_back(line))
   numbers.append(number)
res2 = sum([int(x) for x in numbers])
print(f"Solution 2:\n{res2}")



